nifi-toolkit/nifi-toolkit-encrypt-config/src/main/groovy/org/apache/nifi/properties/ConfigEncryptionTool.groovy

```diff
@@ -810,21 +815,52 @@ class ConfigEncryptionTool {
-        String migratedFlowXmlContent = flowXmlContent.replaceAll(WRAPPED_FLOW_XML_CIPHER_TEXT_REGEX) { String wrappedCipherText ->
-            String plaintext = decryptFlowElement(wrappedCipherText, existingFlowPassword, existingAlgorithm, existingProvider)
-            byte[] cipherBytes = encryptCipher.doFinal(plaintext.bytes)
-            byte[] saltAndCipherBytes = concatByteArrays(encryptionSalt, cipherBytes)
-            elementCount++
-            "enc{${Hex.encodeHex(saltAndCipherBytes)}}"
+        File tempFlowXmlFile = new File(getTemporaryFlowXmlFile(outputFlowXmlPath).toString())
+        BufferedWriter tempFlowXmlWriter = getFileOutputStream(tempFlowXmlFile, isflowXmlGZipped)
+
+        // Scan through XML content as a stream, decrypt and re-encrypt fields with a new flow password
+        final BufferedReader reader = new BufferedReader(new InputStreamReader(flowXmlContent))
+        String line;
+
+        while((line = reader.readLine()) != null) {
+            def matcher = line =~ WRAPPED_FLOW_XML_CIPHER_TEXT_REGEX
+            if(matcher.find()) {
+                String plaintext = decryptFlowElement(matcher.getAt(0), existingFlowPassword, existingAlgorithm, existingProvider)
+                byte[] cipherBytes = encryptCipher.doFinal(plaintext.bytes)
+                byte[] saltAndCipherBytes = concatByteArrays(encryptionSalt, cipherBytes)
+                elementCount++
+                tempFlowXmlWriter.writeLine(line.replaceFirst(WRAPPED_FLOW_XML_CIPHER_TEXT_REGEX, "enc{${Hex.encodeHex(saltAndCipherBytes)}}"))
+            } else {
+                tempFlowXmlWriter.writeLine(line)
+            }
         }
+        tempFlowXmlWriter.flush()
+        tempFlowXmlWriter.close()
+
+        // Overwrite the original flow file with the migrated flow file
+        Files.move(tempFlowXmlFile.toPath(), Paths.get(outputFlowXmlPath), StandardCopyOption.ATOMIC_MOVE)
 
         if (isVerbose) {
             logger.info("Decrypted and re-encrypted ${elementCount} elements for flow.xml.gz")
         }
 
-        migratedFlowXmlContent
+        loadFlowXml(outputFlowXmlPath)
+    }
+
+    private BufferedWriter getFileOutputStream(File outputFlowXmlPath, boolean isFileGZipped) {
+        BufferedWriter migratedFlowXml
+        if(isFileGZipped) {
+            migratedFlowXml = new BufferedWriter(new OutputStreamWriter(new GZIPOutputStream(new FileOutputStream(outputFlowXmlPath))));
+        } else {
+            migratedFlowXml = new BufferedWriter(new OutputStreamWriter(new FileOutputStream(outputFlowXmlPath)));
+        }
+        migratedFlowXml
+    }
+
+    // Create a temporary output file we can write the stream to
+    private Path getTemporaryFlowXmlFile(String originalOutputFlowXmlPath) {
+        String migratedFileName = "migrated-".concat(Paths.get(originalOutputFlowXmlPath).getFileName().toString())
```

Comment:
For improved readability, it may be better to declare a separate variable for the original file path, then declare migratedFileName:
```suggestion
        String migratedFileName = "migrated-${originalFileName}"
```